### ISSUE TITLE
 Router ordering of paths is not respected

### ISSUE THREAD

**Issue by jayvdb** [tfranzel/drf-spectacular#92]
This appears in the now familiar case of [oscarapi](https://github.com/django-oscar/django-oscar-api) and [oscarapicheckout](https://gitlab.com/thelabnyc/django-oscar/django-oscar-api-checkout).

The order in `urls.py` must be (c.f. README of oscarapicheckout):

```py
    url(r'^api/', include(apps.get_app_config("oscarapicheckout").urls[0])), # Must be before oscar_api.urls
    url(r'^api/', include(oscar_api.urls)),
```

However drf-spectacular processes the second one, including the identical path, and internally replaces the first view/serializer/etc with the unused second view/serializer/etc.

If I invert the order, the desired oscarapicheckout endpoint appears in the schema, but then it is *not* the one used by the url router.

**Comment by jayvdb** [comment]
A hacky workaround would be a 'fix' extension which could force drf-spectacular to ignore the second view it encounters.

**Comment by jayvdb** [comment]
Complicating things slightly is an extra `/oscarapi/checkout{format}` which is the `oscar-api` view, which drags in the extra unwanted 'identically named' `CheckoutSerializer` even if the `/oscarapi/checkout` route is fixed, but that is a different problem, afaics.

**Comment by tfranzel** [comment]
very interesting. spectacular should process the routes as they come in. the endpoint sorting might be influencing it though. 

i believe this is an dict override (in spectacular) versus. "take the first matching route" in Django. 

i'll look into it.

**Comment by jayvdb** [comment]
Not a huge priority for me, as I need my hacking of the urlconf to also get rid of `/oscarapi/admin`.  I'm not sure how frequent this over-riding of endpoints happens in Django projects - Oscar is the first time I have seen it done so commonly by plugins.

**Comment by tfranzel** [comment]
@jayvdb please review the change and close. urlpattern overrides should now be properly handled in the schema.

**Comment by jayvdb** [comment]
Sorry I didnt reply earlier.  It is a bit hard for me to easily verify anything as I have diverged so far.  Project almost over.  Will get to tidy/catch up later this week.

**Comment by tfranzel** [comment]
np. you probably also have to wrap your ``oscar-api-checkout `` overrides with https://www.django-rest-framework.org/api-guide/format-suffixes/#format_suffix_patterns in the urlpatterns. that way the override should be "complete".

**Comment by jayvdb** [comment]
Ya, or more likely forcibly filter oscar-api urls to remove the suffix `{format}`, as it needlessly duplicates stuff in the api until https://github.com/tfranzel/drf-spectacular/issues/110 is fixed.

**Comment by jayvdb** [comment]
@tfranzel , I just noticed that this code was duplicated from DRF core, which means the bug probably exists there.  The code there hasnt changed.  There isnt any open issue related to "get_api_endpoints".  https://github.com/encode/django-rest-framework/issues/6527 looks really relevant - but appears closed without a fix.

@Lucidiot , do you know whether your problem was eventually solved in DRF core?

**Comment by Lucidiot** [comment]
My issue occurred while support for OpenAPI was barely starting in DRF 3.9; at that time, the default `--format` was still `coreapi` in `manage.py generateschema`. The gist of the issue was that if you had more than one namespace (if you had at least one `include()` and another include or a path in your `urlpatterns`), all namespaces would be ignored and the OpenAPI schema would have 0 endpoints.

With Django 2.2 and DRF 3.9 I managed to recreate my issue:

``` python
ns1 = [
    path('foo', FooView.as_view(), name='foo')
]

ns2 = [
    path('bar', BarView.as_view(), name='bar')
]

urlpatterns = [
    path('ns1/', include((ns1, 'ns1'), namespace='ns1'))
    path('ns2/', include((ns2, 'ns2'), namespace='ns2'))
]
```

With CoreAPI, I would get the two `/ns1/foo` and `/ns2/bar` as expected, but not with OpenAPI. This bug was indeed fixed with #6532, when actual support for OpenAPI was added. With DRF >=3.10, I do have two endpoints in all schema formats.

But I just tried out something else instead: If I still have my two views with different patterns and different namespaces and different everything, but I use the same serializer in both views, then both views get the same `operationId: RetrieveFoo`, which would make one endpoint override the other anyway with APIStar or other libraries that rely on the `operationId`. Is this what is meant by the `/oscarapi/checkout{format}` overriding the other?
